aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,125 @@
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer. 
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+			size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    int iterations =0;
+    int total_buffer_size= buffer->entry[buffer->in_offs].size;
+    int commands_index = buffer->in_offs;
+   if((buffer == NULL) ) /*Invalid argument check*/
+   {
+   	return NULL;
+   }
+
+
+   while(iterations < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)	/*Running for max iterations permissible*/
+    {
+    
+    	if(char_offset <= (total_buffer_size - 1))			
+        {
+        
+        	*entry_offset_byte_rtn = char_offset-(total_buffer_size - buffer->entry[commands_index].size);
+         	return &buffer->entry[commands_index];
+        }
+          	
+      	else if(char_offset == (total_buffer_size - 1))		/*When offset equals end of one command*/
+      	{
+      		*entry_offset_byte_rtn = char_offset;
+         	return &buffer->entry[commands_index];
+      	
+      	}
+        else 
+        {
+            if(++commands_index == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) /*When buffer is fully traversed*/
+            {
+                commands_index = 0;
+            }
+           
+           total_buffer_size += buffer->entry[commands_index].size;
+        }
+
+        iterations++;
+    }
+
+    return NULL;
+
+}
+
+/**
+* Adds entry @param add_entry to @param buffer in the location specified in buffer->in_offs.
+* If the buffer was already full, overwrites the oldest entry and advances buffer->out_offs to the
+* new start location.
+* Any necessary locking must be handled by the caller
+* Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
+*/
+void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
+{
+    /**
+    * TODO: implement per description 
+    */
+   if(buffer->full == false)						/*Adding command when buffer is not full*/
+   {
+        buffer->entry[buffer->in_offs] = *(add_entry);
+        buffer->in_offs++;
+   }
+   else if(buffer->full == true)					/*Adding command when buffer is full*/
+   {
+   	buffer->entry[buffer->in_offs] = *(add_entry);
+        buffer->out_offs++;
+        buffer->in_offs++; 
+         if(buffer->out_offs == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+  	 {
+        	buffer->out_offs = 0;
```

Comment:
@shreyan-collab think about generic logic using % operator here instead of if check logic